### ISSUE TITLE
Feature request: Hierarchical groups

### ISSUE THREAD

**Issue by PostMidnight** [NodeBB/NodeBB#7410]
Hi Developers.

There are often situations where hierarchical groups are used.
For example, group A containing several sub-groups (B1-n), each containing members (BxM1-n).

```
A -> B1 -> (B1M1 – B1Mn)
        B2 -> (B2M1 – B2Mn)

        Bn -> (BnM1 – BnMn)
```

With hierarchy support, we can assign people to groups B1-Bn and members will be included (automatically) in A.
I am presenting a simple hierarchy (2-level) for sake of simplicity. Things can get much more complex, and thus should be automated (taking care of replications and flagging loops).

Highly appreciated!
JJ.

**Comment by PatheticMustan** [comment]
This would be reeeaally nice, but would make a lot of breaking changes?
(I would love it though :D )

**Comment by PostMidnight** [comment]
@PatheticMustan I would assume the groups will accept children of all types (members and/or subgroups) so this should not be a breaking change.

**Comment by PostMidnight** [comment]
@barisusakli @julianlam 
I spent some time looking at the code.
I believe we can implement hierarchy considering the below changes. Each is minor, however they must be done in multiple places :(
1. Adding GID Integer field to group struct? GID is a unique group identifier.
2. Adding an GIDs integer array field to the group struct: The array will store the child groups (each identified by its integer GID).
3. Adding a GIDs integer array field in the user's struct that will store the groups that the user belongs to (in integer format)? - Here we'll store all the groups the user belongs to, flattening out the hierarchy. 
4. Add a new function isMemberByGID(GIDs[]) - a function that takes an array of GIDs as parameter and returns t/f if the user is a member.

We also need a script that will traverse all groups and assign initial GIDs.

I am deliberately ignoring the GUI required to set the hierarchy at the client - and assuming this functionality will be limited to API.

**Comment by pitaj** [comment]
Consider how this could affect privileges: if privileges are assigned to a group deep in the tree, traversing all the way up could be expensive.

**Comment by PatheticMustan** [comment]
Imagine:
Potterhouse:
    Slytherin
    Hufflepuff
    Ravenclaw
    Gryffindor

How would the tag look like, though? Would it look like "Slytherin", or "Potterhouse>Slytherin"?

**Comment by PostMidnight** [comment]
@pitaj you brought up a very important topic, thank you.

Kindly note, I am *not* suggesting to change the privilege system.

Please allow me to explain: It appears that groups do not retain the privileges: groups are just buckets of members. On the other hand, the categories retains a whitelist set of rules, each having a specific group or member assigned (am I wrong?). Even today one may create two groups containing similar members and assign conflicting rules, but things work, why? because white-list is implemented.

For example, lets assume user X is a member of group A. Lets assume that category Y has the following rules for "edit" privilege:
1. Group-A : allow edit
2. User-X: do not allow edit
The result is allow edit, as each privilege category (in our example: edit permission) is a result of OR of all groups/members that had a rule specified in said category. 

I am suggesting that each user will have a cached flat list of groups he is assigned to (inc. the parent groups) stored together with his account struct. 

So assume the following group tree:
A  -- B
    -- C  -- D
            -- E

Users in E will have the following groups:
E - default in NodeBB
C - Parent (E)
A - Parent (C)

In a similar manner, users in B will have: [B, A]

As for performance, if you will accept my suggestion pertaining group GID (integer), the isMember function will access the user's GIDs array and cycle through it looking for a match to the rule's group. Even a linear search should not take long. If I am not mistaken, using GIDs instead of groupname will probably speed up isMember vs. the current implementation.

Since the user changes groups less often than access, re-building the per-user group flat list has limited performance impact. Admittedly, such change requires going through all affected users.

Bottom line, and again, if my assumptions are not totally off: 
1. No change to the privilege system.
2. Adding few fields to support hierarchical groups is not a breaking change.
3. No changes to tags (@PatheticMustan).

**Comment by pitaj** [comment]
Ok yes I suppose if there is a flat gid array per user, then checking privileges would be the same or faster performance.

**Comment by julianlam** [comment]
I appreciate the work you all have put in, but at this time there are no plans to introduce group hierarchies. Groups in our intention are flat by nature in order to keep its usage simple. Introducing hierarchies would add additional complexity that we are not willing to undertake at this time.

**Comment by PostMidnight** [comment]
I totally understand your position.
If possible, and again, only if you don't mind, might I ask you to consider adding an Unique Integer GID to the group object and an empty integer array. With these two changes I believe I'll be able to get the hierarchy working by myself and then share the code with you. Please don't go out of your way for this. Thanks again.
